### PR TITLE
alter_table: Support adding columns to tables

### PR DESCRIPTION
**Pull request by ParkMyCar**
This PR adds support for `ALTER TABLE ... ADD COLUMN ...` behind two dyncfgs:

1. `enable_alter_table_add_column`
2. `persist_dangerous_enable_schema_evolution`

The Persist dyncfg can be removed once compaction can handle blobs at two different schemas.

The general approach implemented here is to version tables, which builds on top of https://github.com/MaterializeInc/materialize/pull/29647 and https://github.com/MaterializeInc/materialize/pull/29231. The column definitions in a table will have versions attached to them, as well as references to objects, e.g.

```
CREATE TABLE t1 (a int, b text VERSION ADDED 1);

-- Will only contain column 'a' from 't1'
CREATE VIEW v1 AS SELECT * FROM [u1 as "materialize"."public"."t1" VERSION 0];
```

The PR is split into separate commits which can be reviewed independently:

1. Updates to support versioning of `RelationDesc`s
    1. Changes the in-memory `Table` object to use a `VersionedRelationDesc`. This is a wrapper around a `RelationDesc` which provides a few methods for adding columns which increments the existing version, and getting the `RelationDesc` for a specific version of a table.
    2. Updates Catalog APIs to require specifying a version when getting a `RelationDesc` for an object. To make this change in the least invasive way I introduced a new `trait VersionedCatalogEntry: CatalogEntry` and moved the `fn desc(...) -> RelationDesc` method to it. This makes it such that only callers that need a `RelationDesc` have to concern themselves with a version, as opposed to all callers of catalog methods.
    3. Plumbing versions to all necessary spots.
2. Updated planning and sequencing to support adding columns to tables.
    1. This is pretty straight forward, it adds a new `Op::AlterAddColumn` which updates the `create_sql` for a table and adds a new column with `VERSION ADDED <num>` to the table definition.
3. Makes sure when the `enable_alter_table_add_column` feature is not enabled we don't write `VERSION`s for any objects.
    1. Because we also need to add version references in objects that refer to tables (e.g. `VIEW`s), this commit makes sure when the feature is off we don't try to version things. So as-is this PR should result in 0 changes when the feature is off.
4. Adds some validation to `VERSION`s when specified in a column definition or item reference.
5. `bin/fmt`

Also part of all of these commits are updates to `alter-table.slt` to exercise the new functionality.

### Motivation

Fixes: https://github.com/MaterializeInc/database-issues/issues/8233

### Tips for reviewer

It might be easiest to review this PR commit-by-commit

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] If this PR includes major [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note), I have pinged the relevant PM to schedule a changelog post.

### CODE REVIEW

**Comment by def-** [comment]
Tell me when this is ready for testing, there are some tests/frameworks where this will be a good fit (platform-checks, parallel-workload).

File: src/catalog/src/memory/objects.rs [hunk 1]
```diff
@@ -1356,6 +1355,46 @@ impl CatalogItem {
...
+        name: ColumnName,
+        typ: ColumnType,
+        sql: RawDataType,
+    ) -> Result<(), PlanError> {
+        let CatalogItem::Table(table) = self else {
+            return Err(PlanError::Unsupported {
+                feature: "adding columns to a non-Table".to_string(),
+                discussion_no: None,
+            });
+        };
```
Comment (jkosh44) [inline src/catalog/src/memory/objects.rs:1369]: I think this should probably be validated in planning (if it isn't already). I think these functions are usually called after the catalog commit, which means that we may have already committed bad data.

**Review by jkosh44 — APPROVED** [review]
LGTM

**Comment by ParkMyCar** [comment]
Need to iterate more on some changes here

**Comment by ParkMyCar** [comment]
Taking a different approach

**Pull request closed without merging**